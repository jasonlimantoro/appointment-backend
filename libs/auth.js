import jose from 'node-jose';
import { CognitoUser, AuthenticationDetails } from 'amazon-cognito-identity-js';
import config, { userPool } from '../config/aws-exports';
import {
  TokenNotIssuedError,
  PublicKeyNotFoundError,
  ApolloTokenExpiredError,
} from './errors';

class Auth {
  static cognitoUser = null;

  static verifyJwt = async token => {
    const sections = token.split('.');
    // get the kid from the headers prior to verification
    let header = jose.util.base64url.decode(sections[0]);
    header = JSON.parse(header);
    const { kid } = header;
    // download the public keys
    const response = await fetch(config.publicKeysUrl);
    const body = await response.json();
    const { keys } = body;
    // search for the kid in the downloaded public keys
    const correctKey = keys.find(k => k.kid === kid);
    if (!correctKey) {
      throw new PublicKeyNotFoundError();
    }
    // construct the public key
    const toVerify = await jose.JWK.asKey(correctKey);
    // verify the signature
    const verified = await jose.JWS.createVerify(toVerify).verify(token);
    const claims = JSON.parse(verified.payload);
    // additionally we can verify the token expiration
    const currentTs = Math.floor(new Date() / 1000);
    if (currentTs > claims.exp) {
      throw new ApolloTokenExpiredError({
        data: { name: 'TokenExpiredError', claims },
      });
    }
    // and the Audience (use claims.client_id if verifying an access token)
    if (claims.aud !== config.Auth.userPoolWebClientId) {
      throw new TokenNotIssuedError();
    }
    return claims;
  };

  static authenticatePromise = ({ authenticationDetails, cognitoUser } = {}) => new Promise((resolve, reject) => {
    cognitoUser.authenticateUser(authenticationDetails, {
      onSuccess: result => resolve(result),
      onFailure: err => reject(err),
    });
  });

  static getSessionPromise = () => new Promise((resolve, reject) => {
    this.cognitoUser.getSession((err, session) => {
      if (err) reject(err);
      resolve(session);
    });
  });

  static refreshSessionPromise = ({ refreshToken } = {}) => new Promise((resolve, reject) => {
    this.cognitoUser.refreshSession(refreshToken, (err, session) => {
      if (err) reject(err);
      resolve(session);
    });
  });

  static login = async ({ username, password } = {}) => {
    const authenticationDetails = new AuthenticationDetails({
      Username: username,
      Password: password,
    });
    const cognitoUser = new CognitoUser({ Username: username, Pool: userPool });
    const user = await this.authenticatePromise({
      authenticationDetails,
      cognitoUser,
    });
    this.cognitoUser = cognitoUser;
    return user;
  };

  /**
   * @todo: fix issue where cognitoUser is always null
   */
  static logout = () => {
    const cognitoUser = userPool.getCurrentUser();
    cognitoUser.globalSignOut({
      onSuccess: res => console.log(res),
      onFailure: err => console.log(err),
    });
  };

  static refresh = async claim => {
    const userData = {
      Username: claim['cognito:username'] || claim.cognitoUsername,
      Pool: userPool,
    };
    this.cognitoUser = new CognitoUser(userData);
    const session = await this.getSessionPromise();
    const refreshToken = session.getRefreshToken();
    const refreshedSession = await this.refreshSessionPromise({
      userData,
      refreshToken,
    });
    return refreshedSession;
  };
}

export default Auth;
