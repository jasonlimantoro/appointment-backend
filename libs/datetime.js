const pad = n => (n < 10 ? `0${n}` : n);

export const humanFormat = obj => {
  const y = obj.getFullYear();
  const mon = pad(obj.getMonth());
  const d = pad(obj.getDate());
  const h = pad(obj.getHours());
  const min = pad(obj.getMinutes());
  const s = pad(obj.getSeconds());
  return `${y}-${mon}-${d} ${h}:${min}:${s}`;
};
