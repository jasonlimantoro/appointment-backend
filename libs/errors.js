class BaseError extends Error {
  constructor(message) {
    super(message);
    this.name = this.constructor.name;
    Error.captureStackTrace(this, this.constructor);
  }
}
class TokenExpiredError extends BaseError {
  constructor(message = 'Token is expired') {
    super(message);
  }
}

class TokenNotIssuedError extends BaseError {
  constructor(message = 'Token is not issued for this audience') {
    super(message);
  }
}

class PublicKeyNotFoundError extends BaseError {
  constructor(message = 'Public key not found in JWK') {
    super(message);
  }
}

class AuthenticationError extends BaseError {
  constructor(message = 'You are not authenticated') {
    super(message);
  }
}

class TokenVerificationError extends BaseError {}

export {
  PublicKeyNotFoundError,
  TokenExpiredError,
  TokenNotIssuedError,
  TokenVerificationError,
  AuthenticationError,
};
