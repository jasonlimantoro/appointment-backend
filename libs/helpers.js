export const transformObjectKeysToLower = obj => {
  const result = {};
  for (const key in obj) {
    const value = obj[key];
    if (Array.isArray(value)) {
      const transformedArray = [];
      for (let i = 0; i < value.length; i++) {
        const item = value[i];
        if (typeof item === 'object') {
          transformedArray.push(transformObjectKeysToLower(item));
        } else {
          transformedArray.push(item);
        }
      }
      result[key.toLowerCase()] = transformedArray;
    } else if (typeof value === 'object') {
      result[key.toLowerCase()] = transformObjectKeysToLower(value);
    } else {
      result[key.toLowerCase()] = value;
    }
  }
  return result;
};

export const encryptBase64 = str => Buffer.from(str).toString('base64');

export const decryptBase64 = base64 => Buffer.from(base64, 'base64').toString('ascii');
