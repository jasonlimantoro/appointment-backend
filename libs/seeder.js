import faker from 'faker';
import path from 'path';
import { humanFormat } from './datetime';

faker.seed(123);

/**
 * Temporary fix because it is not yet implemented
 */
faker.system.directoryPath = () => path.format({
  base: faker
    .fake('{{random.words}}')
    .replace(/ /g, path.sep)
    .toLowerCase(),
});

/**
 * Temporary fix becuase it is not yet implemented
 */
faker.system.filePath = () => faker.system.directoryPath() + path.sep + faker.system.fileName();

export default class Seeder {
  static seedNumber = {
    guests: 10,
    entries: 15,
    photos: 20,
  };

  static guest = () => ({
    firstName: faker.name.firstName(),
    lastName: faker.name.lastName(),
    NIK: faker.finance.account(16),
    company: faker.company.companyName(),
    email: faker.internet.email(),
  });

  static entry = (guestId, ended = true) => ({
    id: faker.random.uuid(),
    see: faker.name.findName(),
    createdAt: humanFormat(faker.date.past(1, '2019-08-01')),
    guestId,
    endedAt: ended ? humanFormat(faker.date.future(1, '2019-08-01')) : null,
  });

  static photo = (entryId, createdAt) => ({
    id: faker.random.uuid(),
    key: faker.system.filePath(),
    entryId: entryId || faker.random.uuid(),
    createdAt: createdAt || humanFormat(faker.date.past()),
  });
}
export const arrayOf = (times, generator, argsProducer = () => []) => {
  const result = [];
  for (let i = 0; i < times; ++i) {
    result.push(generator(...argsProducer()));
  }

  return result;
};
