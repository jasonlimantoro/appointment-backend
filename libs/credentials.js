import AWS from 'aws-sdk';
import get from 'lodash/get';
import * as helpers from './helpers';

export const getServiceWithAssumedCredentials = async (
  user,
  Service,
  replacerFn,
  sessionName = 'proxiedCognitoRole',
) => {
  const sts = new AWS.STS();
  const role = helpers.selfOrFirstInArray(user['cognito:roles']);
  try {
    const { Credentials } = await sts
      .assumeRole({ RoleArn: role, RoleSessionName: sessionName })
      .promise();
    const ResolvedService = get(AWS, Service);
    const replacedService = new ResolvedService({
      accessKeyId: Credentials.AccessKeyId,
      secretAccessKey: Credentials.SecretAccessKey,
      sessionToken: Credentials.SessionToken,
      ...(process.env.IS_OFFLINE
        && Service.indexOf('DynamoDB') !== -1 && {
        endpoint: 'http://localhost:8000',
        region: 'localhost',
      }),
    });
    if (typeof replacerFn === 'function') {
      replacerFn(replacedService);
      return;
    }
    return replacedService;
  } catch (e) {
    throw e;
  }
};
