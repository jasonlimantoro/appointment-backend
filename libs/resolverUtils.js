import moment from 'moment';
import mime from 'mime-types';
import { getNestedObjectValue } from 'appointment-common';
import Auth from './auth';
import { AuthenticationError } from './errors';
import { transformObjectKeysToLower } from './helpers';
import { humanFormat } from './datetime';

/**
 * Authenticates request using the JWT token in the header
 * @param {object} context the context object sent
 * @param {function} controller the function to be executed if authentication is successful
 * @param  {...any} params arguments for the controller function
 */
export const checkAuthentication = async (context, controller, ...params) => {
  context.headers = transformObjectKeysToLower(context.headers);
  const authorization = getNestedObjectValue(context)([
    'headers',
    'authorization',
  ]);
  if (!authorization) {
    throw new AuthenticationError(
      `Provided header is invalid: ${JSON.stringify(context.headers)}`,
    );
  }
  const token = authorization.split(' ')[1];
  const user = await Auth.verifyJwt(token);
  if (!user) throw new AuthenticationError();
  return controller.apply(this, [...params, context]);
};

export const filterToday = list => {
  const midnight = moment()
    .hours(0)
    .minutes(0);
  const later = moment()
    .hours(23)
    .minutes(59);
  return list.filter(
    ({ createdAt, endedAt }) => !endedAt && moment(createdAt).isBetween(midnight, later),
  );
};

/**
 * To retrieve the lambda request ID
 * @param {object} context the context object sent
 * @returns {string}
 */
export const getLambdaRequestId = context => getNestedObjectValue(context)(['event', 'requestContext', 'requestId']);

/**
 * To retrieve the api request ID
 * @param {object} context the context object sent
 * @returns {string}
 */
export const getAPIRequestId = context => getNestedObjectValue(context)(['context', 'awsRequestId']);

export const constructFileName = ({ prefix = '', fileName, fileType } = {}) => `${prefix}/${humanFormat(new Date())}-${fileName}.${mime.extension(
  fileType,
)}`;
