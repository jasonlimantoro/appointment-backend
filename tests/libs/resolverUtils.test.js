import moment from 'moment';
import { filterToday, checkAuthentication } from '../../libs/resolverUtils';
import Auth from '../../libs/auth';
import { AuthenticationError } from '../../libs/errors';

jest.mock('jsonwebtoken');

describe('resolverUtils', () => {
  describe('filterToday', () => {
    it('should work', () => {
      const given = [
        {
          createdAt: moment().hours(12),
        },
        {
          createdAt: moment().hours(15),
          endedAt: moment().hours(16),
        },
        {
          createdAt: moment().hours(-1),
        },
      ];
      const actual = filterToday(given);
      expect(actual).toEqual([given[0]]);
    });
  });

  describe('checkAuthentication', () => {
    it('should throw error when header is invalid', async () => {
      const controller = jest.fn();
      const context = {};
      await expect(checkAuthentication(context, controller)).rejects.toThrow(
        AuthenticationError,
      );
    });

    it('should throw error when jwt verification fails', async () => {
      const controller = jest.fn();
      Auth.verifyJwt = jest.fn().mockResolvedValue(false);

      const context = {
        headers: {
          authorization: 'Bearer some-token',
        },
      };
      await expect(checkAuthentication(context, controller)).rejects.toThrow(
        AuthenticationError,
      );
      expect(Auth.verifyJwt).toBeCalledWith('some-token');
    });

    it('should work', async () => {
      const controller = jest.fn().mockResolvedValue({ message: 'some-value' });
      Auth.verifyJwt = jest.fn().mockResolvedValue(true);
      const context = {
        headers: {
          authorization: 'Bearer some-token',
        },
      };
      const res = await checkAuthentication(context, controller);
      expect(controller).toBeCalled();
      expect(Auth.verifyJwt).toBeCalledWith('some-token');
      expect(res).toEqual({ message: 'some-value' });
    });
  });
});
