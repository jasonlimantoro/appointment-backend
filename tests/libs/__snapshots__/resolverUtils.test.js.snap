// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolverUtils constructFileName should work 1`] = `"some-file-dir/12:00:00 AM-jason-gunawan.png"`;
